kazoo/client.py

```diff
@@ -640,6 +653,10 @@ def _call(self, request, async_object):
             async_object.set_exception(SessionExpiredError())
             return False
 
+        if self.rate_limiting_sem:
+            self.rate_limiting_sem.acquire(blocking=True)
+            async_object.rawlink(lambda _res: self.rate_limiting_sem.release())
```

Comment:
I saw we have some "recursive" async calls. Is there a way this can deadlock?

I think because we raw link to the Zookeeper response itself, the lambda being called, and then any other links that are used to implement these recursive calls are afterwards?

Reply:
We can probably test this by parametrizing one of the async test suits to accept a KazooClient with the limit set to 1

Reply:
We have in `kazoo.client` some chained async calls (`makepath=True` for example).
I am not sure what you mean by recursive otherwise.

The ZK protocol guarantees that request are always executed in order (no reordering), so I think this is safe no matter what, no?

Reply:
The makepath was what I was thinking, yeah. Struggling to phrase it, but what I was thinking was that we might be sensitive to the order of the callbacks in https://github.com/python-zk/kazoo/blob/master/kazoo/handlers/utils.py#L133. 

Now that I look a bit further, the queue where we add the requests and the queue where the completions of the requests go are separate so I think it's ok.